src/Components/WebAssembly/WebAssembly/src/Hosting/MutableConfiguration.cs

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.Components.WebAssembly.Hosting
+{
+    /// <summary>
+    /// MutableConfiguration is a class that implements the interface of an IConfiguration,
+    /// IConfigurationRoot, and IConfigurationBuilder. It can be used to simulatneously build
+    /// and read from a configuration object.
+    /// </summary>
+    public class MutableConfiguration : IConfiguration, IConfigurationRoot, IConfigurationBuilder
+    {
+        private readonly IList<IConfigurationProvider> _providers = new List<IConfigurationProvider>();
+        private readonly IList<IConfigurationSource> _sources = new List<IConfigurationSource>();
+
+        private readonly IList<IDisposable> _changeTokenRegistrations = new List<IDisposable>();
+        private ConfigurationReloadToken _changeToken = new ConfigurationReloadToken();
+
+        // Lock for setters and getters on the configuration object.
+        private readonly object mutex = new object();
+
+        /// <summary>
+        /// Returns the sources used to obtain configuration values.
+        /// </summary>
+        public IList<IConfigurationSource> Sources => _sources;
+
+        /// <summary>
+        /// Returns the providers user to obtain configuration values.
+        /// </summary>
+        public IEnumerable<IConfigurationProvider> Providers => _providers;
+
+        /// <summary>
+        /// Gets a key/value collection that can be used to share data between the <see cref="IConfigurationBuilder"/>
+        /// and the registered <see cref="IConfigurationProvider"/>s.
+        ///
+        /// In this implementation, this largely exists as a way to satisfy the
+        /// requirements of the IConfigurationBuilder and is not populated by
+        /// <see cref="MutableConfiguration"/> with any meaningful info.
+        /// </summary>
+        public IDictionary<string, object> Properties { get; } = new Dictionary<string, object>();
+
+        /// <summary>
+        /// Gets or sets a configuration value in all providers that are currently
+        /// registered on the <see cref="MutableConfiguration"/> object.
+        /// </summary>
+        /// <param name="key">The configuration key.</param>
+        /// <returns>The configuration value.</returns>
+        public string this[string key] {
+            get {
+                lock (mutex)
```

Comment:
If we're doing synchronization for this, it needs to be in all of the methods that read/write mutable state. So that would include most of the functionality here.

I think you could also make the argument that we don't need to support synchronization because there's no way to do multi-threading in wasm, and also because one shouldn't hold on to this object after startup.

Reply:
There's some future-proofing here. It operates under the assumption that we'll have multi-threading soon and this is one less codepath to have to add support for thread safety to when that point arrives. I can probably remove it from this PR to limit the scope of the changes.